src/serac/coefficients/coefficient_extensions.hpp

```diff
@@ -14,41 +14,170 @@
+#include <vector>
 #include "mfem.hpp"
 
 #include "serac/numerics/expr_template_ops.hpp"
-
+#include "serac/physics/utilities/finite_element_state.hpp"
 /**
  * @brief Functionality that extends current MFEM capabilities
  *
  */
 namespace serac::mfem_ext {
 
+namespace detail {
+// methods for determining index type
+template <typename T>
+struct index_t {
+  using type = std::size_t;
+};
+
+template <typename T>
+struct index_t<mfem::Array<T>> {
+  using type = int;
+};
+
+// methods for determining the size of a container
+template <typename T>
+auto size(T container)
+{
+  return container.size();
+}
+
+template <typename T>
+std::size_t size(mfem::Array<T> container)
+{
+  return container.Size();
+}
+
+// methods for determining type of coefficient evaluations
+template <typename T, typename = void>
+struct eval_t {
+  using type = T;
+};
+
+template <typename T>
+struct eval_t<T, std::enable_if_t<std::is_base_of_v<mfem::Coefficient, T>>> {
+  using type = double;
+};
+
+template <typename T>
+struct eval_t<T, std::enable_if_t<std::is_base_of_v<mfem::VectorCoefficient, T>>> {
+  using type = mfem::Vector;
+};
+
+/// methods for evaluating coefficient stuff with the same prototype
+double eval(double& d, mfem::ElementTransformation&, const mfem::IntegrationPoint&) { return d; }
+
+/// evaluates an mfem::Coefficient
+double eval(mfem::Coefficient& c, mfem::ElementTransformation& Tr, const mfem::IntegrationPoint& ip)
+{
+  return c.Eval(Tr, ip);
+}
+
+/// evaluates a mfem::VectorCoefficient
+mfem::Vector eval(mfem::VectorCoefficient& v, mfem::ElementTransformation& Tr, const mfem::IntegrationPoint& ip)
+{
+  mfem::Vector temp(v.GetVDim());
+  v.Eval(temp, Tr, ip);
+  return temp;
+}
+
+
+template <typename T, typename = void>
+struct is_iterable : std::false_type {};
+  
+template <typename T>
+struct is_iterable<T, std::void_t<decltype(std::declval<T>().begin()),
+                                  decltype(std::declval<T>().end())>>
+    : std::true_type {};
+
+template <typename T, typename = void>
+struct has_brackets : std::false_type {};
+  
+template <typename T>
+struct has_brackets<T, std::void_t<decltype(std::declval<T>()[0])>>
+    : std::true_type {};
+  
+}  // namespace detail
+
+namespace digitize {
 /**
- * @brief MakeTrueEssList takes in a FESpace, a vector coefficient, and produces a list
- *  of essential boundary conditions
+ * @brief takes floating point value and rounds down to the nearest integer
+ * @param[in] v floating point value
+ */
+
+[[maybe_unused]] static int floor(double v) { return static_cast<int>(v); }
+
+/**
+ * @brief Thresholds a real function so that it is 2 if greater than 0.
+ * @param[in] v floating point value
+ */
+[[maybe_unused]] static int threshold(double v) { return v > 0. ? 2 : 1; }
+
+/**
+ * @brief Checks if floating point value is equal to 1, if return 2 otherwise return 1.
+ */
+[[maybe_unused]] static int equals1(double v) { return v == 1. ? 2 : 1; }
+}  // namespace digitize
```

Comment:
It feels like these functions should return `bool` rather than `int`. The convention of "2" for true and "1" for false seems dangerous, as all other C/C++ code assumes "1" for true and "0" for false.

The function name "threshold" is ambiguous too, maybe we could change it to be self-documenting, e.g. `is_positive`, or `greater_than_zero`

Reply:
It just so happens that `threshold` and `equals1` return 1 or 2, but `floor` indicates that you can return a discrete set of integer attributes. It just so happens in these functions the assumption is to have two attributes. Mfem attributes must be `0` because they do something like `array[T.attribute-1]`.

I can add some more comments. These are mostly convenience functions for common "digitization" functions.

I can rename to `greater_than_zero`.